### ISSUE TITLE
Idea: functional API for Pipelines

### ISSUE THREAD

**Issue by maxpumperla** [KonduitAI/konduit-serving#81]
I came across this yesterday: https://github.com/alegonz/baikal

It's essentially the functional API from Keras applied to scikit learn pipelines. i.e. you supply an actual pipeline step to your next step (instead of duplicating output-of-last is input-of-next etc.). For konduit that would mean to establish

- an `Input` abstraction
- `Output`
- the actual chaining of steps

that could be tested with Python first in an experimental version, but could work just as well in Java

**Comment by wmeddie** [comment]
In theory this should be completely doable when you think about it:

`Input` abstraction `InputType`
`Output` abstraction `OutputType`
`Step` becomes `PipelineSteps`

Then instead of a `Model` you create a `Service` which would contain the remaining info we need.  With the Service you could 

1. `start()` it in place
2. `save('config.json')` to save the resulting config.json (magic needed here)
3. `create_server('folder')` to create a konduit-serving bundle with the config.json and all python files that you can just `bin/konduit-server start`
4. `deploy('http://server:1234')` to upload and deploy the pipeline on that server (More magic needed here)

I'm imagining it could look like:

```python
import numpy as np
from keras.models import load_model
from konduit.functional import Input, Output, Service, Step, Top5ClassesStep

def my_op(x, y):
    return x * y / 2

my_model = load_model('foo.h5')

x = Input(dtype=np.array)
y = Input(dtype=float)

op = Step(my_op, output=Output(dtype=np.array)
net = Step(my_model, output=Output(dtype=np.array)


pipeline = op(x, y)
pipeline = net(res)
pipeline = Top5ClassesStep(nclasses=10)(res)

service = Service(pipeline)

service.save("backup_config.json")
service.deploy("http://localhost:1234")
```


Should we not reuse the names Input, Output and Step though?  Since many other things use those names I wonder....

**Comment by maxpumperla** [comment]
yeah, along those lines. following the keras spec, the equivalent design would be to make a `Pipeline` inherit from `PipelineStep`. Then in Python you can implement `__call__` methods on each implementation of `PipelineStep` that return a `PipelineStep` (and take one as input, of course). This allows you to do do roughly this:

```python
from konduit import Input, Output, Pipeline, ModelStep, PythonStep

x = Input(input_name='x', ...)

ps = PythonStep(python_code='x * 2')(x)
ms = ModelStep(type='KERAS', model_file='model.h5)
out = Output(type='RAW')(ms)

server = Server(input=x, output=y, serving_config=...)
```

You can almost do the same thing in Scala with `apply`, in Java we'll have to have a specific `step` method or so to make it work. Actually, I'd like to see this implemented in Java and generate the Python stuff (minus the syntactic sugar).

**Comment by maxpumperla** [comment]
the key thing here is that a lot of internals can be inferred from the former steps. e.g. in the above example, why would a user need to specifically output names for a step and then make sure the input names of the next step exactly match? nobody should care about the internal naming conventions of a pipeline, as long as you can make sure what comes out of it makes sense to you.

**Comment by ShamsUlAzeem** [comment]
So the `ModelStep` would take `ps` from `PythonStep` as `ms = ModelStep(type='KERAS', model_file='model.h5)(ps)`?

**Comment by ShamsUlAzeem** [comment]
Also, maybe we can already have information from the `out` object and just do `server = Server(pipeline=out, serving_config=...)`
